spackbot/workers.py

```diff
@@ -41,23 +45,105 @@ def is_up_to_date(output):
 
     url = job.meta["post_comments_url"]
-    data = {"body": user_msg}
+    data = {"body": msg}
 
     helpers.synchronous_http_request(url, data=data, token=token)
-    logger.error(user_msg)
+    logger.error(msg)
+
+
+def report_style_failure(job, connection, type, value, traceback):
+    user_msg = comments.format_error_message(
+        "I encountered an error attempting to format style.", type, value, traceback)
+    post_failure_message(job, user_msg)
+
+
+def report_rebuild_failure(job, connection, type, value, traceback):
+    user_msg = comments.format_error_message(
+        "I encountered an error attempting to rebuild everything.", type, value, traceback)
+    post_failure_message(job, user_msg)
+
+
+async def run_pipeline_task(event):
+    """
+    Send an api request to gitlab telling it to run a pipeline on the
+    PR branch for the associated PR.  Depending on the job metadata
+    """
+    job = get_current_job()
+    token = job.meta["token"]
+    rebuild_everything = True if "rebuild_everything" in job.meta and job.meta["rebuild_everything"] is True else False
+
+    async with aiohttp.ClientSession() as session:
+        gh = gh_aiohttp.GitHubAPI(session, REQUESTER, oauth_token=token)
+
+        # Get the pull request number
+        pr_url = event.data["issue"]["pull_request"]["url"]
+        number = pr_url.split("/")[-1]
+
+        # We need the pull request branch
+        pr = await gh.getitem(pr_url)
+
+        # Get the sender of the PR - do they have write?
+        sender = event.data["sender"]["login"]
+        repository = event.data["repository"]
+        collaborators_url = repository["collaborators_url"]
+        author = pr["user"]["login"]
+
+        # If it's the PR author, we allow it
+        if author == sender:
+            logger.info(f"Author {author} is requesting a pipeline run.")
+
+        # If they don't have write, we don't allow the command
+        elif not await helpers.found(
+            gh.getitem(collaborators_url, {"collaborator": sender})
+        ):
+            logger.info(f"Not found: {sender}")
+            return f"Sorry {sender}, I cannot do that for you. Only users with write can make this request!"
+
+        # We need the branch name plus number to assemble the GitLab CI
+        branch = pr["head"]["ref"]
+        pr_mirror_key = f"pr{number}_{branch}"
+        branch = urllib.parse.quote_plus(pr_mirror_key)
+
+        if rebuild_everything:
+            logger.info(f"Deleting s3://{helpers.pr_mirror_bucket}/{pr_mirror_key}")
+
+            # Wipe out PR binary mirror contents
+            s3 = boto3.resource("s3")
+            bucket = s3.Bucket(helpers.pr_mirror_bucket)
+            bucket.objects.filter(Prefix=pr_mirror_key).delete()
+
+        url = f"{helpers.gitlab_spack_project_url}/pipeline?ref={branch}"
+
+        if "rebuild_everything" in kwargs and kwargs["rebuild_everything"]:
+            url = f"{url}&variables[SPACK_PRUNE_UNTOUCHED]=False"
```

Comment:
Can you put a comment here explaining what this does?  e.g. (if this is correct):

```suggestion
            # Disabling SPACK_PRUNE_UNTOUCHED tells `spack ci` not to prune build jobs that are already in 
            # the cache index, and it tells generated build jobs not to check the cache, forcing a rebuild.
            url = f"{url}&variables[SPACK_PRUNE_UNTOUCHED]=False"
```

Reply:
This is still just a draft and wasn't quite complete.  It needs to tell spack not to do either kind of pruning (dag pruning and untouched pruning).  But spack didn't support one of those yet.  I have a draft spack [PR](https://github.com/spack/spack/pull/32596) that adds the needed bits.